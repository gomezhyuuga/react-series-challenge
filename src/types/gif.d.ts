export interface IImageObject {
    url: string
    width: number
    height: number
}
export interface IGIF {
    id: string
    slug: string
    url: string
    // Image Objects. See:
    // https://developers.giphy.com/docs/#rendition-guide
    images: {
        original: IImageObject
        fixed_height: IImageObject
    }
}