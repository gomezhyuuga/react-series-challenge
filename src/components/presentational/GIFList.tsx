import * as React from 'react';
import { IGIF } from '../../types/gif';
import styled from 'styled-components';

import FavButton from './FavButton';
import { connect } from 'react-redux';
import { Dispatch, bindActionCreators, Action } from 'redux';
import { toggle_favorite } from '../../actions/gifs';
import { ActionFunctionAny } from 'redux-actions';

interface IGIFListProps {
    gifs: IGIF[],
    favs: IGIF[],
    onToggleFavorite: any
}

const _div = styled.div`
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
`;

class GIFList extends React.Component<IGIFListProps, any> {
    public render() {
        const props = this.props;
        const { gifs, onToggleFavorite, favs } = props;
        return (
        <_div>
            { gifs ?
                gifs.map(gif => {
                    const callback = onToggleFavorite.bind(null, gif);
                    const active = favs[gif.id];
                    return (
                    <div style={{position: 'relative'}} key={gif.id}>
                        <FavButton onAddFavorite={callback} active={active} />
                        <img src={gif.images.fixed_height.url} />
                    </div>);
                })
            :   <h2>No GIFs found!</h2>  }
        </_div>);
    }
}

function mapStateToProps(state: any) {
    return { favs: {...state.favs} }
}
function mapDispatchToProps(dispatch: Dispatch) {
    return bindActionCreators({
        onToggleFavorite: toggle_favorite
    }, dispatch);
}
export default connect(mapStateToProps, mapDispatchToProps)(GIFList);