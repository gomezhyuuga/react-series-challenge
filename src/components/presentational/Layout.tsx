import * as React from 'react';
import styled from 'styled-components';

export const Container = styled.header`
    width: 100%;
    max-width: 960px;
    margin: 0 auto;
    padding: 0 20px;
    display: flex;
    flex-direction: column;
`;
