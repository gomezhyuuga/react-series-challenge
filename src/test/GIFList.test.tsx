import * as React from 'react';
import 'jest-enzyme';
import 'jest-styled-components'
import configureStore from 'redux-mock-store'
import { shallow, mount, ShallowWrapper } from 'enzyme';
import  { GIFList } from '../components/presentational/GIFList';
import { AppState } from '../types';
import * as GIFS from './GIFS.json'

describe('GIFList', () => {
    const mockStore = configureStore();
    const initialState = { trending: GIFS as string[] }
    let store;

    it('renders withouth crashing', () => {
        store = mockStore(initialState);
        const state = store.getState() as AppState;
        const onToggle = jest.fn();
        const wrapper = mount(<GIFList
            onToggleFavorite={onToggle}
            gifs={state.trending}
            favs={state.favs} />);
        expect(wrapper).toMatchSnapshot();
    });

    describe('with empty GIFS', () => {
        it('matches snapshot', () => {
            store = mockStore({});
            const state = store.getState() as AppState;
            const onToggle = jest.fn();
            const wrapper = mount(<GIFList
                onToggleFavorite={onToggle}
                gifs={state.trending}
                favs={state.favs} />);
            expect(wrapper).toMatchSnapshot();
        });
    });
});