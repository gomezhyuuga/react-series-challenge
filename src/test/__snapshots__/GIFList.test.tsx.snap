// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIFList renders withouth crashing 1`] = `
<styled.div>
  <h2>
    No GIFs found!
  </h2>
</styled.div>
`;

exports[`GIFList with empty GIFS matches snapshot 1`] = `<styled.div />`;
